reactos/drivers/usb/usbhub_new/ioctl.c

```diff
@@ -0,0 +1,1444 @@
+                           IN PIRP Irp)
+{
+    PUSBHUB_FDO_EXTENSION HubExtension;
+    PUSBHUB_PORT_DATA PortData;
+    PIO_STACK_LOCATION IoStack;
+    PULONG PortStatus;
+    NTSTATUS Status;
+
+    DPRINT("USBH_PdoIoctlGetPortStatus ... \n");
+
+    HubExtension = PortExtension->HubExtension;
+
+    InterlockedIncrement(&HubExtension->PendingRequestCount);
+
+    KeWaitForSingleObject(&HubExtension->HubSemaphore,
+                          Executive,
+                          KernelMode,
+                          FALSE,
+                          NULL);
+
+    PortData = &HubExtension->PortData[PortExtension->PortNumber - 1];
+
+    Status = USBH_SyncGetPortStatus(HubExtension,
+                                    PortExtension->PortNumber,
+                                    &PortData->PortStatus,
+                                    sizeof(USBHUB_PORT_STATUS));
+
+    IoStack = Irp->Tail.Overlay.CurrentStackLocation;
+    PortStatus = IoStack->Parameters.Others.Argument1;
+
+    *PortStatus = 0;
+
+    if (PortExtension->Common.SelfDevice == PortData->DeviceObject)
+    {
+        if (PortData->PortStatus.UsbPortStatus.Usb20PortStatus.PortEnabledDisabled)
+        {
+            *PortStatus |= USBD_PORT_ENABLED;
+        }
+
+        if (PortData->PortStatus.UsbPortStatus.Usb20PortStatus.CurrentConnectStatus)
+        {
+            *PortStatus |= USBD_PORT_CONNECTED;
+        }
+    }
+
+    KeReleaseSemaphore(&HubExtension->HubSemaphore,
+                       LOW_REALTIME_PRIORITY,
+                       1,
+                       FALSE);
+
+    if (!InterlockedDecrement(&HubExtension->PendingRequestCount))
+    {
+        KeSetEvent(&HubExtension->PendingRequestEvent,
+                   EVENT_INCREMENT,
+                   FALSE);
+    }
+
+    USBH_CompleteIrp(Irp, Status);
+
+    return Status;
+}
+
+VOID
+NTAPI
+USBH_ResetPortWorker(IN PUSBHUB_FDO_EXTENSION HubExtension,
+                     IN PVOID Context)
+{
+    PUSBHUB_RESET_PORT_CONTEXT WorkItemReset;
+    PUSBHUB_PORT_PDO_EXTENSION PortExtension;
+    PUSB_DEVICE_HANDLE DeviceHandle;
+    NTSTATUS Status;
+    USHORT Port;
+
+    DPRINT("USBH_ResetPortWorker ... \n");
+
+    WorkItemReset = Context;
+
+    PortExtension = WorkItemReset->PortExtension;
+
+    if (!HubExtension)
+    {
+        Status = STATUS_UNSUCCESSFUL;
+        goto Exit;
+    }
+
+    InterlockedIncrement(&HubExtension->PendingRequestCount);
+
+    KeWaitForSingleObject(&HubExtension->HubSemaphore,
+                          Executive,
+                          KernelMode,
+                          FALSE,
+                          NULL);
+
+    Port = PortExtension->PortNumber;
+
+    if (PortExtension->Common.SelfDevice == HubExtension->PortData[Port-1].DeviceObject &&
+        (DeviceHandle = PortExtension->DeviceHandle) != 0)
+    {
+        USBD_RemoveDeviceEx(HubExtension,
+                            DeviceHandle,
+                            USBD_MARK_DEVICE_BUSY);
```

Comment:
I think there's just an overlap in naming conventions. The USBD_ functions here don't indicate a relation to usbd.sys -- it is just the naming used for the wrappers around the bus interface functions.